pkg/networkaware/networkoverhead/networkoverhead.go

```diff
@@ -0,0 +1,577 @@
+}
+
+// Filter : evaluate if node can respect maxNetworkCost requirements
+func (no *NetworkOverhead) Filter(ctx context.Context, cycleState *framework.CycleState, pod *v1.Pod, nodeInfo *framework.NodeInfo) *framework.Status {
+
+	klog.V(4).Infof("Filter: pod %q on node %q", pod.GetName(), nodeInfo.Node().Name)
+	if nodeInfo.Node() == nil {
+		return framework.NewStatus(framework.Error, "node not found")
+	}
+
+	// Check if Pod belongs to an App Group
+	agName := networkawareutil.GetPodAppGroupLabel(pod)
+	if len(agName) == 0 { // Return
+		return nil
+	}
+
+	// Get AppGroup CR
+	appGroup, err := findAppGroupNetworkOverhead(agName, no)
+	if err != nil {
+		klog.ErrorS(err, "Error while returning AppGroup")
+		return nil
+	}
+
+	// Get NetworkTopology CR
+	networkTopology, err := findNetworkTopologyNetworkOverhead(no)
+	if err != nil {
+		klog.ErrorS(err, "Error while returning NetworkTopology")
+		return nil
+	}
+
+	klog.V(6).Info("AppGroup CR: ", appGroup.Name)
+	klog.V(6).Info("Network Topology CR: ", networkTopology.Name)
+
+	// Get Dependencies of the given pod
+	dependencyList := networkawareutil.GetDependencyList(pod, appGroup)
+
+	// If the pod has no dependencies, return
+	if dependencyList == nil {
+		return nil
+	}
+
+	// Get pods from lister
+	selector := labels.Set(map[string]string{agv1alpha1.AppGroupLabel: agName}).AsSelector()
+	pods, err := no.podLister.List(selector)
+	if err != nil {
+		klog.ErrorS(err, "Error while returning pods from appGroup, return")
+		return nil
+	}
+
+	// Return if pods are not yet allocated for the AppGroup...
+	if pods == nil {
+		klog.ErrorS(err, "No pods yet allocated, return")
+		return nil
+	}
+
+	// Pods already scheduled: Get Scheduled List (Deployment name, replicaID, hostname)
+	scheduledList := networkawareutil.GetScheduledList(pods)
+	klog.V(6).Info("scheduledList: ", scheduledList)
+
+	// Check if scheduledList is empty...
+	if scheduledList == nil {
+		klog.ErrorS(err, "Scheduled list is empty, return")
+		return nil
+	}
+
+	// Check if bandwidth and network requirements can be met
+	region := networkawareutil.GetNodeRegion(nodeInfo.Node())
+	zone := networkawareutil.GetNodeZone(nodeInfo.Node())
+	klog.V(6).Info("Node Region: ", region)
+	klog.V(6).Info("Node Zone: ", zone)
+
+	// Sort Costs if manual weights were selected
+	sortNetworkTopologyCosts(no, networkTopology)
+
+	// Create map for cost / destinations. Search for requirements faster...
+	var costMap = make(map[networkawareutil.CostKey]int64)
+
+	// Populate Costmap
+	populateCostMap(no, costMap, networkTopology, region, zone)
+	klog.V(6).Info("costMap: ", costMap)
+
+	var numOK int64
+	var numNotOK int64
+	var ok error
+
+	numOK, numNotOK, ok = checkMaxNetworkCostRequirements(scheduledList, dependencyList, nodeInfo, region, zone, costMap, no)
+	if ok != nil {
+		return framework.NewStatus(framework.Error, "pod hostname not found")
+	}
+
+	klog.V(4).Infof("NumNotOk: %v / numOK: %v ", numNotOK, numOK)
+
+	if numNotOK > numOK {
```

Comment:
shouldn't it be a yes or no logic? why compare the quantifies here?

Reply:
Thanks for bringing up this question. In our experiments, we saw that if we filtered out a node based on only 1 unmet dependency, we might get unscheduled pods since the problem becomes harder to solve and available nodes might not be able to accommodate all dependencies. So, we considered quantifies, and we only filter a node if the number of unmet dependencies is higher than the ones that are respected.

For 1 pod dependency, this would fail.

For 10 pod dependencies, filtering a node that respects 9 dependencies makes the scheduling problem hard to solve. That's why our design decision.

Reply:
ok, how about renaming `numOK, numNotOK` to `satisfied, violated`, and adding a comment above `if satisfied > violated` to explain the intent?